### ISSUE TITLE
Add Canvas overlay

### ISSUE THREAD

**Issue by nkh** [nkh/P5-App-Asciio#64]
Canvas calls a function to get a list of positions to overlay with a single character.

**Comment by nkh** [comment]
@qindapao 

I've added an *overlay* branch.

https://github.com/nkh/P5-App-Asciio/branches

The documentation to use it is in the developer's documentation.

You can write a callback that will be called after all the elements are drawn. You can then make your computation but instead for adding filler elements you return a list of fillers characters, and their coordinates, and Asciio will draw them.

**Comment by qindapao** [comment]
@nkh 

Ok, I'll try it when I have time.

**Comment by qindapao** [comment]
'set overlays'   => [ '0A0-o', sub { $_[0]->set_overlays_sub(sub { [0, 0, 'O'], [1, 0, 'K'] }) ; $_[0]->update_display ; } ],

@nkh 

How can I set my own sub here? Can you give me an example? I wrote several kinds on the company computer, but none of them were correct.

**Comment by qindapao** [comment]
My sub is here, App:Asciio::Cross::create_cross_coordinates

**Comment by qindapao** [comment]
'set overlays' => [ '0A0-o', sub { $[0]->set_overlays_sub(sub {\&App:Asciio::Cross::create_cross_coordinates}) ; $[0]->update_display ; } ],  

not right


'set overlays' => [ '0A0-o', sub { $[0]->set_overlays_sub(\&App:Asciio::Cross::create_cross_coordinates) ; $[0]->update_display ; } ],            

also not right

I still don't understand perl deeply.

**Comment by nkh** [comment]
```perl
'set overlays' => # the name of the binding
    [ # binding array contains information about the binding
    '0A0-o', # the shortcut for the binding
    sub # the function that is called when the shortcut is pressed
        {
        # $_, you must use the underscore, is the first argument, $asciio
        $_[0]->set_overlays_sub # call the set_overlays_sub in the $asciio object
                        (
                       \&App:Asciio::Cross::create_cross_coordinates # a reference to a sub called when overlays are needed
                        )  ;
        $[0]->update_display ; # update the display after the shortcut is pressed, might not be needed for you
        }
 ], 
```

**Comment by nkh** [comment]
Did you write this in the file where you define the bindings? at the top.

```
use App::Asciio::Cross ;
```

**Comment by qindapao** [comment]
'set overlays' => [ '0A0-o', sub { $[0]->set_overlays_sub(\&App:Asciio::Cross::create_cross_coordinates) ; $[0]->update_display ; } ],        not working

but


'set overlays' => [ '0A0-o', sub { $[0]->set_overlays_sub(\&App:Asciio::Actions::ElementsManipulation::create_cross_coordinates) ; $[0]->update_display ; } ],          working


i do not know why

**Comment by qindapao** [comment]
Cross is part of the Asciio package.ElementsManipulation Not part of the Asciio package.I only know this.

**Comment by qindapao** [comment]
I know why,


'set overlays' => [ '0A0-o', sub { $[0]->set_overlays_sub(\&App:Asciio::create_cross_coordinates) ; $[0]->update_display ; } ],  working


Cross It's a sub-package

**Comment by nkh** [comment]
hmm, we'll need to thin about Toolfunc.pm (which is a weird name) and Cross.pm, not sure they should be in or out of the Asciio namespace.

**Comment by qindapao** [comment]
That was a major operation,How about we stabilize the current function and then deal with it?

I've run it, and the effect is very good. It seems that the filler can be completely killed.


However, there is an obvious performance degradation. I don't know whether it comes from the calculation coordinate part or the drawing part. I'll take time to look at the time-consuming distribution.

**Comment by nkh** [comment]
***It should be faster not slower***

The computation is the same, except if you changed it.

we replace the creation of all the fillers with a single callback.

The only thing I can think of is that before it wasn't called at every update and now it is.

Commit and I'll look at it.

**Comment by qindapao** [comment]
I can start writing code that removes the cross mode.Then use overlay.
It should start on Wednesday night.
It should be easy, just handle the function of exporting to ascii.

**Comment by nkh** [comment]
Commit what you have, even slow, I'm curious.

Do you want me to write the function exporting to ascii?

**Comment by qindapao** [comment]
@nkh 

I'm still in the company, just leave this part to me.I will deal with it when I have time, and I am very happy to make the cross model truly integrated.

**Comment by nkh** [comment]
It's all yours :)

There may be a way I can generate data you can use more efficiently but we can look at that later.

**Comment by nkh** [comment]
merged in 9b2e337

@qindapao discussion continues in #62
